---
title: "Pre-Imputation Quality Control"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
params:
  project: test-gwas-test
  date:
  build:
  chip: test-chip
  samples:  
  snps: 
  snps_final: 
  samples_merged: 
  samples_final: 
  samples_excluded: 
  snps_excluded: 
  filter_statistics: 
  merged_filter_statistics: 
  maf: 0
  hwe: 1E-6
  snp_call_rate: 0.9
  sample_call_rate: 0.5
  version: 1.0.0
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(warn = -1)

library(ggplot2)
library(DT)
library(dplyr)
```


```{r, echo=FALSE}
samples <- read.csv(params$samples_final, sep="")
samples_count <- length(unique(samples$sample))
snps <- read.csv(params$snps_final, sep="")
snps_count <- length(snps$snp)
if (!file.size(params$samples_excluded) == 0) {
    excluded_samples <- read.csv(params$samples_excluded, sep="",header = FALSE)
    excluded_samples_count <- dim(excluded_samples)[1]
}else {
  excluded_samples_count = 0
}
if (!file.size(params$snps_excluded) == 0) {
    excluded_snps <- read.csv(params$snps_excluded, sep="",header = FALSE)
    excluded_snps_count <- dim(excluded_snps)[1]
}else {
  excluded_snps_count = 0
}

mergedFilterStatistics <- read.csv(params$merged_filter_statistics, sep="")


```

# Summary

| Parameter     | Value                                      |
| ------------- |--------------------------------------------|
| Project       | `r params$project`                         |
| Created on       | `r params$date`                         |
| Pipeline Version | `r  params$version` |
| Chip          | `r params$chip`                            |
| Build          | `r params$build`                            |
| Samples       | `r samples_count - excluded_samples_count` (`r excluded_samples_count` excluded) |
| SNPs          | `r snps_count - excluded_snps_count`                                            |


# Quality Control Steps

```{r, echo=FALSE}
filenames <- unlist(strsplit(params$filter_statistics, " "))
steps <-  read.csv(filenames[1],  sep="")
steps$run = filenames[1]
for (filename in filenames[-1]){
  step <- read.csv(filename, sep="")
  step$run = filename
  steps <- rbind(steps,step)
}

step00 = steps[steps$name == "step00", ]
step01 = steps[steps$name == "step01", ]
step01$filtered_snps = step00$snps - step01$snps
step02 = steps[steps$name == "step02", ]
step02$filtered_snps = step01$snps - step02$snps
step03 = steps[steps$name == "step03", ]
step03$filtered_snps = step02$snps - step03$snps
step04 = steps[steps$name == "step04", ]
step04$filtered_snps = step03$snps - step04$snps
step05 = steps[steps$name == "step05", ]
step05$filtered_snps = step04$snps - step05$snps

#ggplot(steps, aes(x=name, y=snps)) +
#  geom_bar(stat="identity") +
#  facet_grid(. ~ run, ) +
#  xlab("Steps") +
#  ylab("Remaining Variants") +
#  ggtitle("QC Filters - Summary")

```

| Parameter     | Value                                      |
| ------------- |--------------------------------------------|
| Number of Runs       | `r length(filenames)`                            |
| SNPs per Run          | `r prettyNum(mean(step00$snps), scientific=FALSE, big.mark=",")` +/- `r prettyNum(sd(step00$snps), scientific=FALSE, big.mark=",")`                           |

## Raw Data

```{r, echo=FALSE}
datatable(step00)
```

## Step 1: Excluding all indels

Excludes all variants with one or more multi-character allele codes. Variants with single-character allele codes outside of {'A', 'C', 'G', 'T', 'a', 'c', 'g', 't', <missing code>} are also excluded (e.g. indels coded as I or D).

```{r, echo=FALSE}
datatable(step01)
```



## Step 2: Excluding all non-autosomale SNPs

Excludes all variants not listed as autosomal-chromosomes (1-22).

```{r, echo=FALSE}
datatable(step02)
```

## Step 3: Updating strand flips

Updates the chromosome, position and strand of binary ped files using strand files from https://www.well.ox.ac.uk/~wrayner/strand/.

The strand files contain all the variants where the match to the relevant genomic sequence >90%. The strand file contains six columns, SNP id, chromosome, position, %match to genome, strand and TOP alleles.

The strand files assume that your genotype calling algorithm has exported the allele calls aligned to the **Illumina TOP strand** or **Illumina Source strand**. Please check you input file with http://mccarthy.well.ox.ac.uk/chipendium/ui/ to find correct strand files.

```{r, echo=FALSE}
datatable(step03)
```

## Step 4: Excluding all non-autosomale snps (after update strand flips)

Excludes all variants not listed as autosomal-chromosomes (1-22).

```{r, echo=FALSE}
datatable(step04)
```

----

## Step 5: Harmonizing ref/alt alleles and retain only SNPs with known ref-allele

Updates the ref- and alt-allele based on ref-alt files from https://www.well.ox.ac.uk/~wrayner/strand/RefAlt.html. Excludes all SNPs where no ref-allele is known.

```{r, echo=FALSE}
datatable(step05)
```

## Step 6: Merging all single runs into one dataset

```{r, echo=FALSE}
step06 <-  read.csv(params$samples_merged,  sep="")
datatable(step06)
```

## Step 7: Excluding SNPs with low MAF or with extreme HWE deviations

- MAF: `r params$maf` 
- HWE: `r params$hwe` 

```{r, echo=FALSE}
excluded_snps_hwe_maf_count <- step06[1,'snps'] - mergedFilterStatistics[1,'snps']
```

**`r excluded_snps_hwe_maf_count`** snps excluded due low MAF or extreme HWE deviations.

## Step 8: Excluding SNPs with low call rate

**`r excluded_snps_count`** snps excluded due to low call rate (Call rate < `r params$snp_call_rate` ).


## Step 9: Excluding samples with low call rate

**`r excluded_samples_count`** samples excluded due to low call rate (Overall call rate < `r params$sample_call_rate` or at least one 20MB chunk with call rate < `r params$sample_call_rate`).

```{r, echo=FALSE}
if (excluded_samples_count > 0) {
    datatable(excluded_samples)
}
```

# Runs

```{r samples, echo=FALSE, results="asis"}
filenames_samples <- unlist(strsplit(params$samples, " "))
filenames_snps <- unlist(strsplit(params$snps, " "))
for (i in 1:length(filenames)){
  filename_samples <- filenames_samples[i]
  filename_snp <- filenames_snps[i]
  filename <- filenames[i]

  samples <- read.csv(filename_samples, sep="")
  cat("\n")
  cat("<h2>", filename, "</h2>")
  cat("\n")
  print(
    ggplot(samples, aes(x=chunk, y=sample)) +
      geom_tile(aes(fill=call_rate)) +
      scale_fill_gradient(limits=c(0,1)) +
      theme(axis.text.y = element_text(size = 6), axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
      xlab("Chunks") +
      ylab("Samples") +
      ggtitle("Sample Call Rate")
    )
  snps <- read.csv(filename_snp, sep="")
  print(
    ggplot(snps, aes(x=call_rate)) +
      geom_histogram(binwidth = 0.05) +
      xlab("Call Rate") +
      ylab("#SNPs") +
      ggtitle("SNP Call Rate")
    )  
}
```
